image/tarexport/load.go

```diff
@@ -101,7 +104,45 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
 			}
 		}
 
-		for i, diffID := range img.RootFS.DiffIDs {
+		imgConfigStore := mobyDistribution.NewImageConfigStoreFromStore(l.is, l.ds)
+		var deltaBase io.ReadSeeker
+
+		if img.Config != nil {
+			if base, ok := img.Config.Labels["io.resin.delta.base"]; ok {
+				digest, err := digest.Parse(base)
+				if err != nil {
+					return err
+				}
+
+				stream, err := imgConfigStore.GetTarSeekStream(digest)
+				if err != nil {
+					return fmt.Errorf("loading delta base image %v: %w", digest, err)
+				}
+				defer stream.Close()
+
+				deltaBase = stream
+			}
+
+			if targetConfig, ok := img.Config.Labels["io.resin.delta.config"]; ok {
+				digest := digest.FromString(targetConfig)
+				if _, err := l.is.Get(image.ID(digest)); err == nil {
+					outStream.Write([]byte("Target image already exists locally, no need to load it.\n"))
+					return nil
+				}
+
+				config = []byte(targetConfig)
+			}
+		}
+
+		configRootFS, err := imgConfigStore.RootFSFromConfig(config)
+		if err == nil && configRootFS == nil {
+			return errors.New("nil target image root filesystem")
```

Comment:
What is the reasoning for using a mix of `fmt.Errorf`, `outStream.Write`, and `errors.New` in the same function?

Reply:
Interesting question! We use `fmt.Errorf` and `errors.New` to create `error`s and return them. It's the way to say "something wrong happened here and I believe you, dear caller, can handle it better than I can."

We use `outStream.Write`, on the other hand, to say things directly to the human user.

As for why we have two different functions to create `error`s: `fmt.Errorf` simply wraps `errors.New` adding `printf`-like formatting capabilities. It feels inconsistent that they are on different packages, but the Go standard library has placed everything accepting `printf`-like formatting into the `fmt` package. (So... it's consistent in its inconsistency!)

Reply:
Excellent explanation, thank you!